underscore.comparison.islike.js

```diff
@@ -0,0 +1,70 @@
+/*
+* Tests if an object is like another. This means objects should follow the same
+* structure and arrays should contain the same types.
+*
+* E.g.
+*
+* _.islike(
+*   {name: "James", age: 10, hobbies: ["football", "computer games", "baking"]},
+*   {name: "", age: 0, hobbies: [""]}
+* )
+*/
+(function() {
+  // Establish the root object, `window` in the browser, or `require` it on the server.
+  if (typeof exports === 'object') {
+    _ = module.exports = require('underscore');
+  }
+
+    var islike = function(obj, pattern) {
+        if (typeof pattern === "function") {
+            return obj instanceof pattern;
+        }
+
+        if (typeof obj !== typeof pattern) return false;
+            if (pattern instanceof Array && !(obj instanceof Array)) return false;
+
+                var type = typeof pattern;
+
+                if (type == "object") {
+                    if (pattern instanceof Array) {
+                        if (pattern.length > 0) {
+                            var oTypes = _.uniq(_.map(obj, fTypeof));
+                            var pTypes = _.uniq(_.map(pattern, fTypeof));
+                            if (_.difference(oTypes, pTypes).length) {
+                                return false;
+                            }
+                        }
+                    } else { // object
+                        if (pattern.constructor === pattern.constructor.prototype.constructor) {
+                            // for 'simple' objects we enumerate
+                            for (var k in pattern) {
+                                if (!pattern.hasOwnProperty(k)) {
```

Comment:
I'm of the opinion that if people start overriding `hasOwnProperty` then they get what they deserve :)

However, I take your point about the bug. We could also use `_.has` to fix that.

Regarding clarity, it's a bit of an taste thing I feel. Does the underscore style prefer one or the other?

I think the speed is probably negligible, but since `_.keys` does a `for ... in {hasOwn...}` under the covers it can't be faster. Unless I'm missing something there.